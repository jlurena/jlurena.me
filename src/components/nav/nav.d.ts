import { IconProp } from '@fortawesome/fontawesome-svg-core'
import { TabType, NavProps, Tab } from '../tabs/tab'

export type NavTab = {
  icon: IconProp
  tab?: Tab
  type: TabType
  func?: (() => void)
}

export type NavTabs = Record<string, NavTab>

export type NavProps = {
  isMobileNavOpen: boolean
  selectedTab: string
  handleOnTabClick: (tabKey: string) => void
  tabs: NavTabs
}
