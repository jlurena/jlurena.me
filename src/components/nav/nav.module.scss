@use "../../scss/breakpoints";
@use "../../scss/variables";

.mobileNavigationContainer {
  background-color: variables.$primary-color;
  position: absolute;
  top: variables.$mobile-nav-bar-height;
  left: 0;
  right: 0;
  margin-left: auto;
  margin-right: auto;
  width: 50%;
  transition: all 0.2s ease-in-out;
  display: none;
  opacity: 0;

  &.open {
    display:  block;
    opacity: 100;
    z-index: 999;
  }

  div:last-of-type button {
    border-bottom: 0;
  }
}

.navigationContainer {
  display: none;
  position: absolute;
  z-index: 9999;
  text-align: center;
  background-color: variables.$primary-color;
  height: 100%;
  width: 50px;
  flex-direction: column;
  justify-content: center;

  @include breakpoints.screen-sm {
    display: flex;
  }
}

.navigationSlider {
  > div {
    display: flex;
    flex-direction: column;
    border-top: 1px solid variables.$primary-color;

    &:last-child {
      border-bottom: 1px solid variables.$primary-color;
    }

    &.selected button {
      background-color: white;
      color: variables.$primary-color;
    }

    &:last-of-type {
      margin-right: 0;
    }
  }
}
