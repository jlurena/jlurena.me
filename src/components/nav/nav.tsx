import React from 'react'
import styles from './nav.module.scss'
import NavButton from '../nav-button'
import type { NavProps } from './nav.d'

const Nav: React.FC<NavProps> = ({
  isMobileNavOpen,
  selectedTab,
  handleOnTabClick,
  tabs
}) => {
  const navOptions = Object.keys(tabs).map(k => (
    <NavButton
      ariaLabel={k}
      icon={tabs[k].icon}
      onClick={() => handleOnTabClick(k)}
      key={k}
      labelHelper={k}
      permanentSelect={k === selectedTab}
    />
  ))

  return (
    <>
      <div 
        className={styles.navigationContainer}
        role="navigation"
      >
        <div className={styles.navigationSlider}>
          {navOptions}
        </div>
      </div>
      <div
        className={`${styles.mobileNavigationContainer} ${isMobileNavOpen && styles.open}`}
        role="navigation"
        aria-hidden={!isMobileNavOpen}
      >
        {isMobileNavOpen && navOptions}
      </div>
    </>
  )
}

export default Nav
