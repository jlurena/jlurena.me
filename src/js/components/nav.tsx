import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/fontawesome-common-types';

import styles from '../../scss/components/nav.module.scss';

type IconProps = {
  ariaLabel: string,
  icon: IconDefinition,
  onClick: () => void
};

export const Icon: React.FC<IconProps> = ({ ariaLabel, icon, onClick }) => (
  <button type="button" aria-label={ariaLabel} className={styles.tab} onClick={onClick}>
    <FontAwesomeIcon icon={icon} size="lg" />
  </button>
);

export declare type NavProps = {
  isMobileNav: boolean,
  onTabClick: (tab: string) => void,
  selectedTab: string,
  tabs: { [key: string]: { icon: IconDefinition, tab: React.FC } }
};

const Nav: React.FC<NavProps> = ({
  isMobileNav, selectedTab, onTabClick, tabs
}) => {
  const navOptions = Object.keys(tabs).map(k => (
    <div className={`${styles.tabWrapper} ${selectedTab === k ? styles.selected : ''}`} key={k}>
      <Icon ariaLabel={k} icon={tabs[k].icon} onClick={() => onTabClick(k)} />
      <em>{k}</em>
    </div>
  ));

  return (
    <>
      <div className={styles.navigationContainer}>
        <div className={styles.navigationSlider}>
          {navOptions}
        </div>
      </div>
      <div className={styles.mobileNavigationContainer} style={{ display: isMobileNav ? 'block' : 'none' }}>
        {navOptions}
      </div>
    </>
  );
};

export default Nav;
