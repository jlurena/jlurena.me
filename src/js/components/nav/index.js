import Nav, { Icon } from './nav';

export default Nav;

export { Icon };
