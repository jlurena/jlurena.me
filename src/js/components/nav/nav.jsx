import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import styles from './nav.module.scss';

export const Icon = ({ ariaLabel, icon, onClick }) => (
  <button type="button" aria-label={ariaLabel} className={styles.tab} onClick={onClick}>
    <FontAwesomeIcon icon={icon} size="lg" />
  </button>
);

const Nav = ({
  isMobileNav, selectedTab, onTabClick, tabs,
}) => {
  const navOptions = Object.keys(tabs).map(k => (
    <div className={`${styles.tabWrapper} ${selectedTab === k ? styles.selected : ''}`} key={k}>
      <Icon ariaLabel={k} icon={tabs[k].icon} onClick={() => onTabClick(k)} />
      <em className={styles.helperLabel}>{k}</em>
    </div>
  ));

  return (
    <>
      <div className={styles.navigationContainer}>
        <div className={styles.navigationSlider}>
          {navOptions}
        </div>
      </div>
      <div
        className={styles.mobileNavigationContainer}
        style={
          {
            visibility: isMobileNav ? 'visible' : 'hidden',
            opacity:    isMobileNav ? '100' : '0',
          }
}
      >
        {navOptions}
      </div>
    </>
  );
};

export default Nav;
