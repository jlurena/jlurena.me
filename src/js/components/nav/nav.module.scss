@import "../../../scss/breakpoints";
@import "../../../scss/variables";

.tab {
  cursor: pointer;
  color: white;
  font-size: 17px;
  line-height: 46px;
  padding: 0;
  border: 0;
  height: 46px;
  width: 50px;
  background-color: transparent;
  transition: all 0.3s ease-in-out;

  & ~ em {
    visibility: hidden;
    opacity: 0;
    position: absolute;
    background-color: $primary-color;
    color: white;
    font-size: 12px;
    font-weight: 600;
    left: calc(100% + 45px);
    min-width: 90px;
    padding: 14px 0;
    text-transform: uppercase;
    transform: translateX(-50%);
    transition: opacity 0.2s 0s, visibility 0s 0.2s, transform 0.2s 0s;
    white-space: nowrap;
  }

  &:hover {
    background-color: white;
    color: $primary-color;
    transition: all 0.3s ease-in-out;

    & ~ em {
      opacity: 1;
      visibility: visible;
      transform: translateX(-40%);
      transition: opacity 0.2s 0s, visibility 0s 0s, transform 0.2s 0s;
    }
  }
}

.mobileNavigationContainer {
  background-color: $primary-color;
  position: absolute;
  top: $mobileNavBarHeight;
  left: 0;
  right: 0;
  margin-left: auto;
  margin-right: auto;
  width: 50%;
  transition: all 0.2s ease-in-out;

  button.tab {
    margin-left: auto;
    margin-right: auto;
    width: 100%;
    border-bottom: 0.5px solid rgba(#fff, 0.2);

    ~ em {
      display: none;
    }
  }

  div:last-of-type button.tab {
    border-bottom: 0;
  }
}

.navigationContainer {
  display: none;
  position: absolute;
  z-index: 999;
  text-align: center;
  background-color: $primary-color;
  height: 100%;
  width: 50px;
  flex-direction: column;
  justify-content: center;

  @include screen-sm {
    display: flex;
  }
}

.navigationSlider {
  div {
    display: flex;
    border-top: 1px solid $primary-color;

    &:last-child {
      border-bottom: 1px solid $primary-color;
    }

    &.selected .tab {
      background-color: white;
      color: $primary-color;
    }

    &:last-of-type {
      margin-right: 0;
    }
  }

  .tabWrapper {
    display: flex;
    flex-direction: column;
    justify-content: center;
    border-top: 1px solid color.adjust($primary-color, 5%);

    &:first-of-type {
      border-top: 0;
    }
  }
}

.helperLabel {
  font-style: normal;
}
