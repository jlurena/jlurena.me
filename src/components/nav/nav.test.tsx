import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { faHome, faAd } from '@fortawesome/free-solid-svg-icons';
import Nav from './nav';
import { NavTabs } from './nav.d';
import { NAV_CONTENT_TYPE } from '../tabs/tab.d';

describe('<Nav />', () => {
  function FooComponent() {
    return <div>Hello Foo!</div>;
  }

  function BarComponent() {
    return <div>Hello Bar!</div>;
  }

  const tabs: NavTabs = {
    Foo: { icon: faHome, tab: FooComponent, type: NAV_CONTENT_TYPE },
    Bar: { icon: faAd, tab: BarComponent, type: NAV_CONTENT_TYPE },
  };

  it('renders mobile view and calls onTabClick', () => {
    const onTabClickSpy = vi.fn();

    const { asFragment } = render(
      <Nav
        isMobileNavOpen
        handleOnTabClick={onTabClickSpy}
        selectedTab="Home"
        tabs={tabs}
      />,
    );

    expect(asFragment()).toMatchSnapshot();

    // Trigger first tab button click
    const buttons = screen.getAllByRole('button');
    fireEvent.click(buttons[0]);

    expect(onTabClickSpy).toHaveBeenCalledWith('Foo');
  });

  it('renders desktop view (snapshot)', () => {
    const { asFragment } = render(
      <Nav
        isMobileNavOpen={false}
        handleOnTabClick={vi.fn()}
        selectedTab="Bar"
        tabs={tabs}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
