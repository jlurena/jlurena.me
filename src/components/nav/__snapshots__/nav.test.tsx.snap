// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Nav /> > renders desktop view (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="_navigationContainer_9d65e5"
    role="navigation"
  >
    <div
      class="_navigationSlider_9d65e5"
    >
      <div
        class="_tabWrapper_a7072b null "
      >
        <div
          class="_buttonContainer_a7072b"
        >
          <button
            aria-label="Foo"
            class="_tab_a7072b"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-house fa-lg "
              data-icon="house"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M575.8 255.5c0 18-15 32.1-32 32.1l-32 0 .7 160.2c0 2.7-.2 5.4-.5 8.1l0 16.2c0 22.1-17.9 40-40 40l-16 0c-1.1 0-2.2 0-3.3-.1c-1.4 .1-2.8 .1-4.2 .1L416 512l-24 0c-22.1 0-40-17.9-40-40l0-24 0-64c0-17.7-14.3-32-32-32l-64 0c-17.7 0-32 14.3-32 32l0 64 0 24c0 22.1-17.9 40-40 40l-24 0-31.9 0c-1.5 0-3-.1-4.5-.2c-1.2 .1-2.4 .2-3.6 .2l-16 0c-22.1 0-40-17.9-40-40l0-112c0-.9 0-1.9 .1-2.8l0-69.7-32 0c-18 0-32-14-32-32.1c0-9 3-17 10-24L266.4 8c7-7 15-8 22-8s15 2 21 7L564.8 231.5c8 7 12 15 11 24z"
                fill="currentColor"
              />
            </svg>
          </button>
          <span
            class="_helperLabel_a7072b"
          >
            Foo
          </span>
        </div>
      </div>
      <div
        class="_tabWrapper_a7072b null _selected_a7072b"
      >
        <div
          class="_buttonContainer_a7072b"
        >
          <button
            aria-label="Bar"
            class="_tab_a7072b"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-rectangle-ad fa-lg "
              data-icon="rectangle-ad"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M64 32C28.7 32 0 60.7 0 96L0 416c0 35.3 28.7 64 64 64l448 0c35.3 0 64-28.7 64-64l0-320c0-35.3-28.7-64-64-64L64 32zM229.5 173.3l72 144c5.9 11.9 1.1 26.3-10.7 32.2s-26.3 1.1-32.2-10.7L253.2 328l-90.3 0-5.4 10.7c-5.9 11.9-20.3 16.7-32.2 10.7s-16.7-20.3-10.7-32.2l72-144c4.1-8.1 12.4-13.3 21.5-13.3s17.4 5.1 21.5 13.3zM208 237.7L186.8 280l42.3 0L208 237.7zM392 256a24 24 0 1 0 0 48 24 24 0 1 0 0-48zm24-43.9l0-28.1c0-13.3 10.7-24 24-24s24 10.7 24 24l0 96 0 48c0 13.3-10.7 24-24 24c-6.6 0-12.6-2.7-17-7c-9.4 4.5-19.9 7-31 7c-39.8 0-72-32.2-72-72s32.2-72 72-72c8.4 0 16.5 1.4 24 4.1z"
                fill="currentColor"
              />
            </svg>
          </button>
          <span
            class="_helperLabel_a7072b"
          >
            Bar
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    class="_mobileNavigationContainer_9d65e5 false"
    role="navigation"
  />
</DocumentFragment>
`;

exports[`<Nav /> > renders mobile view and calls onTabClick 1`] = `
<DocumentFragment>
  <div
    class="_navigationContainer_9d65e5"
    role="navigation"
  >
    <div
      class="_navigationSlider_9d65e5"
    >
      <div
        class="_tabWrapper_a7072b null "
      >
        <div
          class="_buttonContainer_a7072b"
        >
          <button
            aria-label="Foo"
            class="_tab_a7072b"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-house fa-lg "
              data-icon="house"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M575.8 255.5c0 18-15 32.1-32 32.1l-32 0 .7 160.2c0 2.7-.2 5.4-.5 8.1l0 16.2c0 22.1-17.9 40-40 40l-16 0c-1.1 0-2.2 0-3.3-.1c-1.4 .1-2.8 .1-4.2 .1L416 512l-24 0c-22.1 0-40-17.9-40-40l0-24 0-64c0-17.7-14.3-32-32-32l-64 0c-17.7 0-32 14.3-32 32l0 64 0 24c0 22.1-17.9 40-40 40l-24 0-31.9 0c-1.5 0-3-.1-4.5-.2c-1.2 .1-2.4 .2-3.6 .2l-16 0c-22.1 0-40-17.9-40-40l0-112c0-.9 0-1.9 .1-2.8l0-69.7-32 0c-18 0-32-14-32-32.1c0-9 3-17 10-24L266.4 8c7-7 15-8 22-8s15 2 21 7L564.8 231.5c8 7 12 15 11 24z"
                fill="currentColor"
              />
            </svg>
          </button>
          <span
            class="_helperLabel_a7072b"
          >
            Foo
          </span>
        </div>
      </div>
      <div
        class="_tabWrapper_a7072b null "
      >
        <div
          class="_buttonContainer_a7072b"
        >
          <button
            aria-label="Bar"
            class="_tab_a7072b"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-rectangle-ad fa-lg "
              data-icon="rectangle-ad"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M64 32C28.7 32 0 60.7 0 96L0 416c0 35.3 28.7 64 64 64l448 0c35.3 0 64-28.7 64-64l0-320c0-35.3-28.7-64-64-64L64 32zM229.5 173.3l72 144c5.9 11.9 1.1 26.3-10.7 32.2s-26.3 1.1-32.2-10.7L253.2 328l-90.3 0-5.4 10.7c-5.9 11.9-20.3 16.7-32.2 10.7s-16.7-20.3-10.7-32.2l72-144c4.1-8.1 12.4-13.3 21.5-13.3s17.4 5.1 21.5 13.3zM208 237.7L186.8 280l42.3 0L208 237.7zM392 256a24 24 0 1 0 0 48 24 24 0 1 0 0-48zm24-43.9l0-28.1c0-13.3 10.7-24 24-24s24 10.7 24 24l0 96 0 48c0 13.3-10.7 24-24 24c-6.6 0-12.6-2.7-17-7c-9.4 4.5-19.9 7-31 7c-39.8 0-72-32.2-72-72s32.2-72 72-72c8.4 0 16.5 1.4 24 4.1z"
                fill="currentColor"
              />
            </svg>
          </button>
          <span
            class="_helperLabel_a7072b"
          >
            Bar
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="false"
    class="_mobileNavigationContainer_9d65e5 _open_9d65e5"
    role="navigation"
  >
    <div
      class="_tabWrapper_a7072b null "
    >
      <div
        class="_buttonContainer_a7072b"
      >
        <button
          aria-label="Foo"
          class="_tab_a7072b"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-house fa-lg "
            data-icon="house"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M575.8 255.5c0 18-15 32.1-32 32.1l-32 0 .7 160.2c0 2.7-.2 5.4-.5 8.1l0 16.2c0 22.1-17.9 40-40 40l-16 0c-1.1 0-2.2 0-3.3-.1c-1.4 .1-2.8 .1-4.2 .1L416 512l-24 0c-22.1 0-40-17.9-40-40l0-24 0-64c0-17.7-14.3-32-32-32l-64 0c-17.7 0-32 14.3-32 32l0 64 0 24c0 22.1-17.9 40-40 40l-24 0-31.9 0c-1.5 0-3-.1-4.5-.2c-1.2 .1-2.4 .2-3.6 .2l-16 0c-22.1 0-40-17.9-40-40l0-112c0-.9 0-1.9 .1-2.8l0-69.7-32 0c-18 0-32-14-32-32.1c0-9 3-17 10-24L266.4 8c7-7 15-8 22-8s15 2 21 7L564.8 231.5c8 7 12 15 11 24z"
              fill="currentColor"
            />
          </svg>
        </button>
        <span
          class="_helperLabel_a7072b"
        >
          Foo
        </span>
      </div>
    </div>
    <div
      class="_tabWrapper_a7072b null "
    >
      <div
        class="_buttonContainer_a7072b"
      >
        <button
          aria-label="Bar"
          class="_tab_a7072b"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-rectangle-ad fa-lg "
            data-icon="rectangle-ad"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M64 32C28.7 32 0 60.7 0 96L0 416c0 35.3 28.7 64 64 64l448 0c35.3 0 64-28.7 64-64l0-320c0-35.3-28.7-64-64-64L64 32zM229.5 173.3l72 144c5.9 11.9 1.1 26.3-10.7 32.2s-26.3 1.1-32.2-10.7L253.2 328l-90.3 0-5.4 10.7c-5.9 11.9-20.3 16.7-32.2 10.7s-16.7-20.3-10.7-32.2l72-144c4.1-8.1 12.4-13.3 21.5-13.3s17.4 5.1 21.5 13.3zM208 237.7L186.8 280l42.3 0L208 237.7zM392 256a24 24 0 1 0 0 48 24 24 0 1 0 0-48zm24-43.9l0-28.1c0-13.3 10.7-24 24-24s24 10.7 24 24l0 96 0 48c0 13.3-10.7 24-24 24c-6.6 0-12.6-2.7-17-7c-9.4 4.5-19.9 7-31 7c-39.8 0-72-32.2-72-72s32.2-72 72-72c8.4 0 16.5 1.4 24 4.1z"
              fill="currentColor"
            />
          </svg>
        </button>
        <span
          class="_helperLabel_a7072b"
        >
          Bar
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
